contracts/src/gateway/router/CheckpointingFacet.sol

```diff
@@ -52,9 +52,42 @@ contract CheckpointingFacet is GatewayActorModifiers {
         bytes32 membershipRootHash,
         uint256 membershipWeight
     ) external systemActorOnly {
-        if (checkpoint.blockHeight % s.bottomUpCheckPeriod != 0) {
-            revert InvalidCheckpointEpoch();
-        }
+        // TODO: Discussion:
+        //
+        // This check is removed:
+        //
+        // if (checkpoint.blockHeight % s.bottomUpCheckPeriod != 0) {
+        //    revert InvalidCheckpointEpoch();
+        // }
+        //
+        // First of all, the check is buggy in the first place as it does not support
+        // a full message batch before a checkpoint period is reached.
+        // Then, `prevBottomUpCheckpointHeight` is introduced, as long as BU checkpoints
+        // can form a chain, it should be fine.
+        //
+        // Instead, one can add the following check:
+        //
+        // if (checkpoint.prevHeight != s.prevBottomUpCheckpointHeight) {
+        //     revert InvalidCheckpointEpoch();
+        // }
+        //
+        // However, we are restricting to ONLY sequential creation
+        // of bottom up checkpoints. One can only submit the next one if the current
+        // checkpoint quorum is reached. This would slow down the creation process.
```

Comment:
You are saying the relayers can only submit checkpoints with full quorum. I think that's reasonable, a checkpoint without a quorum cannot be executed, so what's the point of it? Either there is another checkpoint missing in front of it, or it's an equivocation. The relayer has the infromation to reconstruct the right sequence.

Reply:
Yeah, relayer should submit sequential. Also `fendermint` node will create bottom up checkpoint and immediately sets the `lastBottomUpCheckpointHeight` and this will allow future checkpoints to be `created`. Signature collection will be done later.